### ISSUE TITLE
Bug in Segmentation of reductions with empty tensors?

### ISSUE THREAD

**Issue by kevinstephano** [NVIDIA/Fuser#1270]
There appears to be a segfault while copying a Fusion object.  I think this is for segmentation as there happens to be an inner and outer reduction.  If I remove one or the other, the fusion is just fine.

Repro:

```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1], contiguity=[True, None], dtype=DataType.BFloat16, is_cpu=False)
    T1 = fd.define_tensor(shape=[-1, -1], contiguity=[None, True], dtype=DataType.BFloat16, is_cpu=False)
    T2 = fd.ops.cast(T1, dtype=DataType.Float)
    S3 = fd.define_scalar(1.00000, dtype=DataType.Double)
    T4 = fd.ops.full(fill_value=S3, shape=[5, 0], dtype=DataType.BFloat16)
    T5 = fd.ops.cast(T4, dtype=DataType.Float)
    T6 = fd.ops.mul(T2, T5)
    T7 = fd.ops.cast(T0, dtype=DataType.Float)
    T8 = fd.ops.mul(T7, T5)
    T24 = fd.ops.sum(T6, axes=[1], keepdim=False, dtype=DataType.Null)
    T11 = fd.ops.sum(T8, axes=[0], keepdim=False, dtype=DataType.Null)
    fd.add_output(T24)
    fd.add_output(T11)

inputs = [
         torch.randn(5, 0, device='cuda', dtype=torch.bfloat16),
         torch.randn(5, 0, device='cuda', dtype=torch.bfloat16),
        ]

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

out = fd.execute(inputs)
```

**Comment by jacobhinkle** [comment]
This appears to be a problem with empty tensor removal, maybe related to #1246. Even prior to concretization, we have replaced `T5` with `T7`:
```
Fusion before concretization:                                                                                                                   
Inputs:                                                                                                                                         
  T0_g[ iS0{i0}, bS1{i1} ], __bfloat                                                                                                            
  T1_g[ bS2{i4}, iS3{i5} ], __bfloat                                                                                                            
Outputs:                                                                                                                                        
  T8_g[ iS16{5} ], float                                                                                                                        
  T9_g[ rS17{5}, iS18{0} ], float                                                                                                               
                                                                                                                                                
%kernel_math {                                                                                                                                  
T8_g[ iS16{5} ]                                                                                                                                 
   = full({5}, float(0));                                                                                                                       
T6_l[ iS12{i0}, bS13{i1} ]                                              
   = __bfloat2float(T0_g[ iS0{i0}, bS1{i1} ]);                                                                                                  
f12 = __double2bfloat(double(1));                                       
T3_l[ iS6{5}, iS7{0} ]                                                  
   = full({5, 0}, ( __double2bfloat(double(1)) ));                                                                                              
T4_l[ iS8{5}, iS9{0} ]                                                                                                                          
   = __bfloat2float(T3_l[ iS6{5}, iS7{0} ]);                            
T7_l[ iS14{5}, iS15{0} ]                                                                                                                        
   = T6_l[ iS12{i0}, bS13{i1} ]                                         
   * T4_l[ iS8{5}, iS9{0} ];                                                                                                                    
T9_g[ rS17{5}, iS18{0} ]                                                
   = reduction( T7_l[ iS14{5}, iS15{0} ], op = add, initial value = float(0), allreduce = false )                                               
}
```
But the Fusion still holds `T5` and its definition:
```
T5_l[ iS10{5}, iS11{0} ]                              
   = T2_l[ bS4{i4}, iS5{i5} ]
   * T4_l[ iS8{5}, iS9{0} ];  
```
The `EmptyTensorRemover` pass rightly removes `T4` but does not know it has a "dead use" since `T5` is unused.

I am not sure where `T5` went yet, but that is probably the key to understanding this issue.

**Comment by jacobhinkle** [comment]
Ah, that's it. The sum over a definition-time zero-size axis will result in a `full()`. So the inputs become dead at definition. That is why `T2_l` and `T5_l` are both dead code even before concretization. I think I know how to clean this up in `DeadCodeRemover`.

**Comment by wujingyue** [comment]
Thanks for investigating this, Jacob! Do you want to take the bug over since this is related to your #1246 and you have more context on this?

**Comment by wujingyue** [comment]
Fwiw, the reproducer seems to run fine at head: https://github.com/NVIDIA/Fuser/blob/fix1270/repro.py. I'm not sure how to reproduce this any more...

**Comment by jacobhinkle** [comment]
It is still segfaulting for me on main a0624cb13. Fix in #1280
